uvicorn/protocols/websockets/websockets_sansio_impl.py

```diff
@@ -0,0 +1,389 @@
+        self.request: Request
+        self.response: Response
+        self.curr_msg_data_type: str
+
+        self.read_paused = False
+        self.writable = asyncio.Event()
+        self.writable.set()
+
+        # Buffers
+        self.bytes: "bytes" = b""
+
+    def connection_made(self, transport: BaseTransport) -> None:
+        """Called when a connection is made."""
+        transport = typing.cast(Transport, transport)
+        self.connections.add(self)
+        self.transport = transport
+        self.server = get_local_addr(transport)
+        self.client = get_remote_addr(transport)
+        self.scheme = "wss" if is_ssl(transport) else "ws"
+
+        if self.logger.level <= TRACE_LOG_LEVEL:
+            prefix = "%s:%d - " % self.client if self.client else ""
+            self.logger.log(TRACE_LOG_LEVEL, "%sWebSocket connection made", prefix)
+
+    def connection_lost(self, exc: typing.Optional[Exception]) -> None:
+        self.connections.remove(self)
+        if self.logger.level <= TRACE_LOG_LEVEL:
+            prefix = "%s:%d - " % self.client if self.client else ""
+            self.logger.log(TRACE_LOG_LEVEL, "%sWebSocket connection lost", prefix)
+        if self.handshake_initiated and not self.close_sent:
+            self.queue.put_nowait({"type": "websocket.disconnect", "code": 1006})
+
+    def data_received(self, data: bytes) -> None:
+        try:
+            self.conn.receive_data(data)
+        except Exception:
+            self.logger.exception("Exception in ASGI server")
+            self.transport.close()
+        self.handle_events()
+
+    def shutdown(self) -> None:
+        if not self.transport.is_closing():
+            if self.handshake_complete:
+                self.queue.put_nowait({"type": "websocket.disconnect", "code": 1012})
+                self.close_sent = True
+                self.conn.send_close(1012)
+                output = self.conn.data_to_send()
+                self.transport.writelines(output)
+            elif self.handshake_initiated:
+                self.send_500_response()
+                self.queue.put_nowait({"type": "websocket.disconnect", "code": 1006})
+            self.transport.close()
+
+    def handle_events(self) -> None:
+        for event in self.conn.events_received():
+            if isinstance(event, Request):
+                self.handle_connect(event)
+            if isinstance(event, Frame):
+                if event.opcode == websockets.frames.Opcode.CONT:
+                    self.handle_cont(event)
+                elif event.opcode == websockets.frames.Opcode.TEXT:
+                    self.handle_text(event)
+                elif event.opcode == websockets.frames.Opcode.BINARY:
+                    self.handle_bytes(event)
+                elif event.opcode == websockets.frames.Opcode.PING:
+                    self.handle_ping(event)
+                elif event.opcode == websockets.frames.Opcode.PONG:
+                    self.handle_pong(event)
```

Comment:
as per this document - https://datatracker.ietf.org/doc/html/rfc6455#autoid-29
[5.5.3] Pong
 The Pong frame contains an opcode of 0xA.
   A Pong frame sent in response to a Ping frame must have identical
   "Application data" as found in the message body of the Ping frame
   being replied to.
   If an endpoint receives a Ping frame and has not yet sent Pong
   frame(s) in response to previous Ping frame(s), the endpoint MAY
   elect to send a Pong frame for only the most recently processed Ping
   frame.
   A Pong frame MAY be sent unsolicited.  This serves as a
   unidirectional heartbeat.  A response to an unsolicited Pong frame is
   not expected.
   
   So, the websockets sansI/O layer chooses not to propagate the received PONG frame event.
   That's why It is not being covered.